client/src/app/worker/autoupdate-stream-pool.ts

```diff
@@ -0,0 +1,243 @@
+    public removeStream(stream: AutoupdateStream) {
+        for (let subscription of stream.subscriptions) {
+            if (this.subscriptions[subscription.id]) {
+                delete this.subscriptions[subscription.id];
+            }
+        }
+
+        const idx = this.streams.indexOf(stream);
+        if (idx !== -1) {
+            this.streams.splice(idx, 1);
+        }
+    }
+
+    /**
+     * Updates the given endpoint for all managed streams
+     *
+     * @param endpoint The new endpoint configuration
+     */
+    public setEndpoint(endpoint: { url?: string; healthUrl?: string; method?: string; authToken?: string }) {
+        this.endpoint = Object.assign(this.endpoint, endpoint);
+
+        for (let stream of this.streams) {
+            stream.updateEndpoint(this.endpoint);
+        }
+    }
+
+    /**
+     * @param subscriptionId Id of the subscription
+     *
+     * @returns subscription with given id or null
+     */
+    public getSubscriptionById(subscriptionId: number): AutoupdateSubscription | null {
+        return this.subscriptions[subscriptionId] || null;
+    }
+
+    /**
+     * Searches a subscription that fulfills the given queryParams
+     * and modelRequest
+     *
+     * @param queryParams
+     * @param modelRequest
+     */
+    public getMatchingSubscription(queryParams: string, modelRequest: Object): AutoupdateSubscription | null {
+        for (let stream of this.streams) {
+            for (let subscription of stream.subscriptions) {
+                if (subscription.fulfills(queryParams, modelRequest)) {
+                    return subscription;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private async connectStream(stream: AutoupdateStream, force?: boolean): Promise<void> {
+        const { stopReason, error } = await stream.start(force);
+
+        if (stopReason === `unused`) {
+            this.removeStream(stream);
+        } else if (stopReason === `error`) {
+            await this.handleError(stream, error);
+        } else if (stopReason === `resolved`) {
+            await this.handleError(stream, undefined);
+        }
+    }
+
+    private async isEndpointHealthy(): Promise<boolean> {
+        const data = await fetch(this.endpoint.healthUrl).then(response => {
+            if (response.ok) {
+                return response.json();
+            }
+
+            return { healthy: false };
+        });
+
+        return !!data?.healthy;
+    }
+
+    private sendToAll(action: string, content?: any) {
+        const ports: MessagePort[] = [];
+        for (const stream of this.streams) {
+            for (const subscription of stream.subscriptions) {
+                for (const port of subscription.ports) {
+                    if (ports.indexOf(port) === -1) {
+                        ports.push(port);
+                    }
+                }
+            }
+        }
+
+        for (const port of ports) {
+            port.postMessage({
```

Comment:
Just put this in the above loop as well

something like:
```
            ...
            ports.push(port)
            port.postMessage({...})
            ...
```

No need to iterate over everything _again_.

Also: The deeply nested for loops make this a bit unreadable, maybe create a helper function.

Also: If you'd prefer not to move postMessage into the big for loop, maybe you could turn the ports-Array into a set?
That way you could skip the explicit duplication check and just use `.add(port)`